### PR TITLE
Refine py_reader exit

### PR DESCRIPTION
**Pull request by sneaxiy**
`py_reader` would raise `EOFException` when the user-defined reader raises exceptions. The proper way is to terminate the training process when the user-defined reader raises exceptions. This PR refines the exit way.

### CODE REVIEW

File: paddle/fluid/pybind/reader_py.cc [hunk 1]
```diff
@@ -139,19 +144,44 @@ class MultiDeviceFeedReader {
...
+        try {
+          readers_[i]->ReadNext(&ret_[i]);
+          return ret_[i].empty() ? Status::kEOF : Status::kSuccess;
+        } catch (...) {
+          exceptions_[i] = std::current_exception();
+          return Status::kException;
+        }
       });
     }
   }
```
Comment (zhhsplendid) [inline paddle/fluid/pybind/reader_py.cc:156]: Is it return for a void function?

Comment (sneaxiy) [reply paddle/fluid/pybind/reader_py.cc:156]: No. The return value is captured in `futures_[i]` but not the return value of `ReadAsync`.

Comment (zhhsplendid) [reply paddle/fluid/pybind/reader_py.cc:156]: Sorry, I was reading incorrectly :-)

**Review by zhhsplendid — APPROVED** [review]
LGTM

**Pull request merged**